types/bluebird/index.d.ts

```diff
@@ -659,39 +670,39 @@ declare class Bluebird<R> implements PromiseLike<R>, Bluebird.Inspection<R> {
   // TODO: After https://github.com/Microsoft/TypeScript/issues/2983 is implemented, we can use
   // the return type propagation of generators to automatically infer the return type T.
   static coroutine<T>(
-      generatorFunction: () => IterableIterator<any>,
+      generatorFunction: () => IterableIterator<T>,
```

Comment:
I don't this change is alright yet until https://github.com/Microsoft/TypeScript/issues/2983 is fixed.

Let's assume someone's code currently looks like this:
```ts
Bluebird.coroutine<number>(function *() {
    yield "string";
    return 1234;
});
```
then this would technically be correct: The Promise returned by `Bluebird.coroutine` contains the last yielded value of the `generatorFunction`.

But with your change this will fail to compile!
You can try this yourself if you:
* Go [here](https://github.com/segayuu/DefinitelyTyped/blob/54a1709291ef7d7e6eecab1dbda2cd5a771305b6/types/bluebird/bluebird-tests.ts#L894) and remove the type cast
* Go [here](https://github.com/segayuu/DefinitelyTyped/blob/54a1709291ef7d7e6eecab1dbda2cd5a771305b6/types/bluebird/bluebird-tests.ts#L893) and set the type `T` to `Bar` (i.e. `<Bar, number, string>`)

In the latter case the compile error will be
```
Argument of type '(a: number, b: string) => IterableIterator<Foo | Bar>' is not assignable to parameter of type '(a1: number, a2: string) => IterableIterator<Bar>'.
  Type 'IterableIterator<Foo | Bar>' is not assignable to type 'IterableIterator<Bar>'.
    Type 'Foo | Bar' is not assignable to type 'Bar'.
      Type 'Foo' is not assignable to type 'Bar'.
        Property 'bar' is missing in type 'Foo'.
```

Reply:
I understand that this part is breaking change.
This is also a change according to the [`common-mistakes`](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes) guidelines.
I think that depending on the version of Typescript, appropriate type arguments will change so that they should not depend on it.
However, I think that the current situation that some kind of cast must be inserted is not beautiful. (It is certain that the specific setting of the type argument is producing type casting effect.)

Reply:
I'm not sure if I understand your comment…

Either way: If you can't convince me otherwise I'm fully against this change right now.
Bluebird is one of the most widely used typings and we really need to prevent breaking changes if possible.
As such let's please wait until https://github.com/Microsoft/TypeScript/issues/2983 is fixed.
Afterwards we can implement a fully proper solution here, which correctly infers the type `T` and won't break anything.

Here's another case where your typings won't work:
```ts
function foo(): Bluebird<number> {
    return Bluebird.coroutine(function *() {
        yield "string1";
        yield "string2";
        yield "string3";
        return 1234;
    });
}
```
Using the current typings this will compile, while with yours a explicit `as Bluebird<number>` will be necessary. And if you need `as Bluebird<number>` I don't see yet how that would be any safer than using `IterableIterator<any>`.

Reply:
In other words, I received that it must give priority to maintaining compatibility (as opposed to dtslint) when it comes to libraries as big as bluebird.

I can not deny that opinion, so add tests and rewind.
However, dtslint invalidation is done not by inline but by tslint.json.

The reason is that I do not think that it is "settled" in the current type definition.
Also, by upgrading to version 3.0, automatic type definition of overload will increase more, but this decision will hamper these advantages to some extent.